core/src/main/java/me/lokka30/treasury/plugin/core/utils/PluginDownloader.java

```diff
@@ -0,0 +1,161 @@
+        try {
+            // first get the latest build
+            URL latestBuild = new URL(LATEST_BUILD);
+            try (Reader in = new InputStreamReader(latestBuild.openStream())) {
+                JsonObject object = Utils.GSON.fromJson(in, JsonObject.class);
+                JsonObject build = object.getAsJsonArray("builds").get(0).getAsJsonObject();
+                String downloadUrlCopy = DOWNLOAD_URL;
+                downloadUrlCopy = downloadUrlCopy.replace("%number%",
+                        Integer.toString(build.get("number").getAsInt())
+                );
+
+                String downloadPath = null;
+                for (JsonElement element : build.getAsJsonArray("artifacts")) {
+                    JsonObject artifact = element.getAsJsonObject();
+                    String relativePath = artifact.get("relativePath").getAsString();
+                    if (relativePath.contains(platformName) && !relativePath.contains("-sources")) {
+                        downloadPath = relativePath;
+                        break;
+                    }
+                }
+                if (downloadPath == null) {
+                    // nag the user to nag us that something went wrong
+                    plugin.logger().warn(
+                            "Something went wrong whilst parsing latest build download data. Please notify a Treasury developer ASAP.");
+                    // and return
+                    return;
+                }
+                downloadUrlCopy = downloadUrlCopy.concat(downloadPath);
+
+                // find old jar because after we download we may not be able to find it
+                File oldJar = null;
+                for (File file : plugin
+                        .getPluginsFolder()
+                        .toFile()
+                        .listFiles((dir, name) -> name.contains("treasury"))) {
+                    if (file != null) {
+                        // first not null we find should be the jar
+                        oldJar = file;
+                        break;
+                    }
+                }
+
+                // prepare for a download
+                String[] downloadPathParts = downloadPath.split("/");
+                String fileName = downloadPathParts[downloadPathParts.length - 1];
+                File file = new File(plugin.getPluginsFolder().toFile(), fileName);
+
+                // now download it
+                URL downloadUrl = new URL(URLEncoder.encode(downloadUrlCopy, "UTF-8"));
+                try (InputStream is = downloadUrl.openStream()) {
+                    file.createNewFile();
+                    try (OutputStream out = new FileOutputStream(file)) {
+                        out.write(readAllBytes(is));
+                    }
+                }
+
+                // delete the old jar on exit (if we found it)
+                if (oldJar != null) {
+                    oldJar.deleteOnExit();
+                }
```

Comment:
Since this command runs while the server is running you can't remove the old plugin jar as its being used, or am I mistaken?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/io/File.html#deleteOnExit()

The java.io.File.deleteOnExit() method **deletes** the file or directory defined by the abstract path name **when the virtual machine terminates**. Files or directories are deleted in the reverse order as they are registered.